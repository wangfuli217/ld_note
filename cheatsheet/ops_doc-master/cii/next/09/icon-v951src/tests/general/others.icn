
procedure spell(n)
   local m
   n := integer(n) | stop(image(n)," is not an integer")
   if n <= 12 then return {
      "0zero,1one,2two,3three,4four,5five,6six,7seven,8eight,_
         9nine,10ten,11eleven,12twelve," ? {
            tab(find(n))
            move(*n)
            tab(upto(","))
            }
      }
   else if n <= 19 then return {
      spell(n[2] || "0") ?
         (if ="for" then "four" else tab(find("ty"))) || "teen"
      }
   else if n <= 99 then return {
      "2twen,3thir,4for,5fif,6six,7seven,8eigh,9nine," ? {
         tab(upto(n[1]))
         move(1)
         tab(upto(",")) || "ty" ||
            if n[2] ~= 0 then "-" || spell(n[2])
         }
      }
   else if n <= 999 then return {
      spell(n[1]) || " hundred" ||
         (if (m := n[2:0]) ~= 0 then " and " || spell(m) else "")
      }
   else if n <= 999999 then return {
      spell(n[1:-3]) || " thousand" ||
         (if (m := n[2:0]) ~= 0 then " and " || spell(m) else "")
      }
   else if n <= 999999999 then return {
      spell(n[1:-6]) || " million" ||
         (if (m := n[2:0]) ~= 0 then " and " || spell(m) else "")
      }
   else fail
end

procedure spellw(n)
   write(n, "	", spell(n))
   return
end

procedure main()
   every spellw(1 to 25)
   every spellw(30 to 110 by 3)
   spellw(945123342)
   every spellw(10000000 to 10000500 by 7)
   sieve()
   wordcnt()
end

#
#          S I E V E   O F   E R A T O S T H E N E S
#

#  This program illustrates the use of sets in implementing the
#  classical sieve algorithm for computing prime numbers.

procedure sieve()
   local limit, s, i
   limit := 100
   s := set()
   every insert(s,1 to limit)
   every member(s,i := 2 to limit) do
      every delete(s,i + i to limit by i)
   delete(s,1)
   primes := sort(s)
   write("There are ",*primes," primes in the first ",limit," integers.")
   write("The primes are:")
   every write(right(!primes,*limit + 1))
end

#
#          W O R D   C O U N T I N G
#

#  This program tabulates the words in standard input and writes the
#  results with the words in a column 20 characters wide.  The definition
#  of a "word" is naive.

procedure wordcnt()
   wordcount(20)
end

procedure wordcount(n)
   local t, line, x, i
   static letters
   initial letters := &lcase ++ &ucase
   t := table(0)
   while line := read() do
      line ? while tab(upto(letters)) do
         t[tab(many(letters))] +:= 1
   x := sort(t,3)
   every i := 1 to *x - 1 by 2 do
      write(left(x[i],n),x[i + 1])
end
