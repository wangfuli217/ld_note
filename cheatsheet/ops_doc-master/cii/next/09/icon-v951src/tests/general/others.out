1	one
2	two
3	three
4	four
5	five
6	six
7	seven
8	eight
9	nine
10	ten
11	eleven
12	twelve
21	twenty-one
22	twenty-two
23	twenty-three
24	twenty-four
25	twenty-five
33	thirty-three
36	thirty-six
39	thirty-nine
42	forty-two
45	forty-five
48	forty-eight
51	fifty-one
54	fifty-four
57	fifty-seven
63	sixty-three
66	sixty-six
69	sixty-nine
72	seventy-two
75	seventy-five
78	seventy-eight
81	eighty-one
84	eighty-four
87	eighty-seven
93	ninety-three
96	ninety-six
99	ninety-nine
102	one hundred and two
105	one hundred and five
108	one hundred and eight
945123342	nine hundred and forty-five million and forty-five million and five million and one hundred and twenty-three thousand and twenty-three thousand and three thousand and three hundred and forty-two
10000000	ten million
10000007	ten million and seven
10000021	ten million and twenty-one
10000028	ten million and twenty-eight
10000035	ten million and thirty-five
10000042	ten million and forty-two
10000049	ten million and forty-nine
10000056	ten million and fifty-six
10000063	ten million and sixty-three
10000077	ten million and seventy-seven
10000084	ten million and eighty-four
10000091	ten million and ninety-one
10000098	ten million and ninety-eight
10000105	ten million and one hundred and five
10000112	ten million and one hundred and twelve
10000126	ten million and one hundred and twenty-six
10000133	ten million and one hundred and thirty-three
10000147	ten million and one hundred and forty-seven
10000154	ten million and one hundred and fifty-four
10000161	ten million and one hundred and sixty-one
10000168	ten million and one hundred and sixty-eight
10000175	ten million and one hundred and seventy-five
10000182	ten million and one hundred and eighty-two
10000189	ten million and one hundred and eighty-nine
10000196	ten million and one hundred and ninety-six
10000203	ten million and two hundred and three
10000210	ten million and two hundred and ten
10000224	ten million and two hundred and twenty-four
10000231	ten million and two hundred and thirty-one
10000238	ten million and two hundred and thirty-eight
10000245	ten million and two hundred and forty-five
10000252	ten million and two hundred and fifty-two
10000259	ten million and two hundred and fifty-nine
10000266	ten million and two hundred and sixty-six
10000273	ten million and two hundred and seventy-three
10000287	ten million and two hundred and eighty-seven
10000294	ten million and two hundred and ninety-four
10000301	ten million and three hundred and one
10000308	ten million and three hundred and eight
10000322	ten million and three hundred and twenty-two
10000329	ten million and three hundred and twenty-nine
10000336	ten million and three hundred and thirty-six
10000343	ten million and three hundred and forty-three
10000357	ten million and three hundred and fifty-seven
10000364	ten million and three hundred and sixty-four
10000371	ten million and three hundred and seventy-one
10000378	ten million and three hundred and seventy-eight
10000385	ten million and three hundred and eighty-five
10000392	ten million and three hundred and ninety-two
10000399	ten million and three hundred and ninety-nine
10000406	ten million and four hundred and six
10000427	ten million and four hundred and twenty-seven
10000434	ten million and four hundred and thirty-four
10000441	ten million and four hundred and forty-one
10000448	ten million and four hundred and forty-eight
10000455	ten million and four hundred and fifty-five
10000462	ten million and four hundred and sixty-two
10000469	ten million and four hundred and sixty-nine
10000476	ten million and four hundred and seventy-six
10000483	ten million and four hundred and eighty-three
10000497	ten million and four hundred and ninety-seven
There are 25 primes in the first 100 integers.
The primes are:
   2
   3
   5
   7
  11
  13
  17
  19
  23
  29
  31
  37
  41
  43
  47
  53
  59
  61
  67
  71
  73
  79
  83
  89
  97
A                   2
B                   1
D                   1
I                   1
L                   1
N                   1
O                   2
R                   1
T                   2
This                1
U                   1
W                   1
a                   1
and                 1
characters          1
column              1
do                  3
end                 2
every               2
in                  1
initial             1
input               2
lcase               1
left                1
letters             4
line                3
local               1
main                2
many                1
n                   2
procedure           3
processes           1
results             1
scan                1
sort                1
standard            1
static              1
t                   4
tab                 2
table               1
the                 2
ucase               1
upto                1
using               1
while               1
wide                1
with                1
wordcount           2
words               1
write               1
writes              1
x                   3
y                   4
